feat: add autocopy functionality to CopyToClipboard (#10328)

michael-s-molina opened the pull request:
### SUMMARY
- Adds autocopy functionality to `CopyToClipboard` (#10328)
- Fixes a bug where the copying was being called twice
- Changes `copyTextToClipboard` in `copy.ts` to be asynchronous and fixed promise type
- Adjusts the margin of buttons that contain an icon but no text
- Changes copy feedback to use notifications instead of tooltip changes

I left the copy button in autocopy mode just in case the user navigates to another tab, copy some content, comes back to the application and wants to copy the link again without reopening the modal or popover.

[Alert PR](https://github.com/apache/superset/pull/12122) improves notification positioning with right and bottom margins helping the user to notice the message.

@junlincc @rusackas 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<img width="1773" alt="Screen Shot 2021-02-09 at 4 17 48 PM" src="https://user-images.githubusercontent.com/70410625/107416856-b8812280-6af3-11eb-898f-67e0ff62a3b3.png">
<img width="1782" alt="Screen Shot 2021-02-09 at 4 28 09 PM" src="https://user-images.githubusercontent.com/70410625/107416940-d77fb480-6af3-11eb-9e68-3e08a665f19b.png">
<img width="1772" alt="Screen Shot 2021-02-09 at 4 21 32 PM" src="https://user-images.githubusercontent.com/70410625/107416976-e0708600-6af3-11eb-9556-8151fe6a6cf8.png">
<img width="1787" alt="Screen Shot 2021-02-09 at 4 23 29 PM" src="https://user-images.githubusercontent.com/70410625/107416996-e6666700-6af3-11eb-84e8-7f9cb32385bb.png">

### TEST PLAN
1 - Go to a dashboard
2- Click on a chart 3 dots icon
3 - Select share chart
4 - The link should be automatically copied to clipboard

1 - Go to explore
2 - Click on chain icon in top right corner 
3 - The link should be automatically copied to clipboard

### ADDITIONAL INFORMATION
- [x] Has associated issue
- [x] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

michael-s-molina committed:
Migrates copy URL link and share by email functionalities from modal to one-click solution (#10328)

rusackas commented on superset-frontend/src/components/Button/index.tsx:
```diff
@@ -144,6 +144,17 @@ export default function Button(props: ButtonProps) {
+  } else {
+    renderedChildren = Children.toArray(children);
+  }
+
```
```suggestion
const renderedChildren = Children.toArray(element?.type == React.Fragment ? element.props.children : children);
```
Not sure if it's easier to read in a one liner and/or with a null-coalescing operator. ¯\\\_(ツ)_/¯

rusackas commented:
This looks like some great engineering work @michael-s-molina , and I'm giving it a review. Meanwhile, just thinking about the feature, I have a *small* UX concern/curiosity for @zuzana-vej and/or @junlincc that I feel I should raise.

Were flows around the email function considered, where the user might have something on the clipboard (e.g. the contents of an email) that gets nuked by this automatic copying? While it saves a click, I hope we're not frustrating other users even more.

I'm wondering if that does potential UX harm by overwriting the contents of the clipboard when there's no clear indication that the "share" button will do so. I suspect that's why most sites (Youtube as one example) don't do this.

If the intent of the change is simply to remove a click, then I wonder if Share should trigger a modal in the first place, which requires a click to close. The design of the Share button on Reddit, for example, serves the same purpose - open the menu, click "copy" and it just goes away unobtrusively.

rusackas reviewed (approved):
Minor nit/question, but generally looks good to me from a code perspective, so I'm approving.

Before merging, I just want to make sure this is the approach we want to take on the UI/UX front in general though, as there may be other ways to save a click (or two) without surprising the user in an impactful way.

ktmud commented:
I share the same concern with @rusackas on the implicitness of the auto-copy behavior and how it may accidentally override users' existing clipboard. 

I'm wondering if we can replace the share button with a copy icon or move the share functionality (under two links, "Copy sharable URL" and "Share via Email") under the hamburger menu.

I'd imagine the frustration mentioned in #10328 mostly come from the "slow" animation of the popover and the fact that you have to move your mouse horizontally to the left to find on the Copy icon. Putting the copy link in a dropdown menu might make the action flow smoother.

zuzana-vej commented:
@rusackas Good point, haven't thought about that. There were few users who provided feedback that they need to copy twice just to share a link. Sharing a link is very common and action users perform many times even every day, while the share via email is not - but just because our users don't user it much doesn't mean others in community don't, and I can see how this could be confusing in general. 

@ktmud generally the feedback was that user needs 2 clicks, and that the copy button is small and user needs to navigate to it. Which on dashboard means moving mouse very far. On Chart Explore not but still user needs to move to the left to navigate to the small icon. Displaying the option to copy link RIGHT BELLOW the share link button (and then the email option below that) could be much better user experience (still 2 clicks, but much easier), which also addresses @rusackas's concern. The other option is to have two icons as @ktmud suggested too.

nytai commented:
+1 on this concern. I think there's a reason why the copy to clipboard is a separate action than share for almost all instances where this functionality appears. Wiping the user's clipboard and adding a new item is something that should only happen with the user's consent. I don't think "share" implies this consent.

michael-s-molina commented:
Thank you all for the valuable feedback. Nice catch @rusackas! I really like @ktmud's solution with fewer clicks and seamless integration with the interface. I even simulated this behavior bellow:

<img width="1775" alt="Screen Shot 2021-02-10 at 8 12 35 AM" src="https://user-images.githubusercontent.com/70410625/107503459-be6d1700-6b78-11eb-8c91-6cbc8645c447.png">
<img width="1770" alt="Screen Shot 2021-02-10 at 8 15 36 AM" src="https://user-images.githubusercontent.com/70410625/107503466-c2009e00-6b78-11eb-9470-c453eab3420d.png">

What do you all think?

ktmud commented:
@michael-s-molina Thanks for the quick simulation! 

I think this works for the dashboard page. For the Explore page though, looking at this more, I'm now no so sure about replacing the icon as it's not very clear what does a copy icon mean. Most first-time users would probably associate it with copying the data, the query, or the chart metadata, not a short-url.

Here's another proposal: wow about we keep the popover but make it open on hover and remove the animation?

michael-s-molina commented:
> @michael-s-molina Thanks for the quick simulation!
> 
> I think this works for the dashboard page. For the Explore page though, looking at this more, I'm now no so sure about replacing the icon as it's not very clear what does a copy icon mean. Most first-time users would probably associate it with copying the data, the query, or the chart metadata, not a short-url.
> 
> Here's another proposal: wow about we keep the popover but make it open on hover and remove the animation?

@ktmud I totally agree about the copy icon but I really like the idea to remove the popover and make this action more straightforward. Can't we just use the Fontawesome link icon? It's already being used in other copy buttons like this one:

<img width="121" alt="Screen Shot 2021-02-10 at 4 01 52 PM" src="https://user-images.githubusercontent.com/70410625/107558340-6a365700-6bb9-11eb-909a-7f8a8cce1735.png">

I think the user will quickly assimilate, and also we have the tooltip to help. It could would look like this:

<img width="280" alt="Screen Shot 2021-02-10 at 4 24 17 PM" src="https://user-images.githubusercontent.com/70410625/107560708-766fe380-6bbc-11eb-9afe-a0c670c4caf2.png">

or

<img width="257" alt="Screen Shot 2021-02-10 at 4 21 57 PM" src="https://user-images.githubusercontent.com/70410625/107560496-314bb180-6bbc-11eb-896d-1a3e95e3868c.png">

What do you think?

ktmud commented:
I think the link icon works, too, but let's make sure to add a tooltip "Copy chart URL to clipboard". It'll change the behavior of an existing icon, but maybe that won't be a big issue with the added tooltip?

What do you all think?  @evans @junlincc @nytai @zuzana-vej ?

zuzana-vej commented:
I think the 2nd image (link icon and email icon) is good. Only thing is that dashboard and explore have slightly different behavior (one has two icons and one has another menu item in the dropdown). But I think it's probably ok and I think it will save users time. In terms of the behavior (on Explore) changing, I think that's very straightfoward for users to get used to.

junlincc commented:
We have plan of moving those buttons and share functionalities under menu, leaving only `copy link` directly accessible from the view. let's keep discussion of this temporary solution light. Signing off from Product and Design. Thanks so much @michael-s-molina !

michael-s-molina commented:
In this last commit I did the following changes in addition to original changes:

- Changed icons to use Icon component except for JSON and CSV because we don’t have equivalent icons yet. Maybe with Diego's PR (https://github.com/apache/superset/pull/12229) we can replace these in the future.
- When clicked copy and email buttons were generating a different short URL even when the main URL had no change. I fixed this to avoid unnecessary POST requests.

- Added tooltip for all action buttons in Explore

![hover](https://user-images.githubusercontent.com/70410625/107971112-cd950000-6f90-11eb-9db7-f883de87dbd7.gif)


- Removed outline border when active

<img width="294" alt="Screen Shot 2021-02-12 at 9 14 22 AM" src="https://user-images.githubusercontent.com/70410625/107971147-db4a8580-6f90-11eb-9d7a-19cdc6bd4c0e.png">

- Changed dashboard copy/email from modal to menu item

<img width="1767" alt="Screen Shot 2021-02-15 at 8 52 26 AM" src="https://user-images.githubusercontent.com/70410625/107971839-befb1880-6f91-11eb-9963-cf3e3c6f7758.png">

- Changed chart copy/email from modal to menu item

<img width="573" alt="Screen Shot 2021-02-15 at 8 52 10 AM" src="https://user-images.githubusercontent.com/70410625/107971589-6f1c5180-6f91-11eb-8ed5-dc37d86fc666.png">


I also created a React hook called `useUrlShortener` for the cases where we need a short version of an URL.

ktmud commented:
This is so cool. Thanks for all the updates and the attention to details! @michael-s-molina

ktmud commented on superset-frontend/src/explore/components/ExploreActionButtons.jsx:
```diff
@@ -115,6 +163,19 @@ export default function ExploreActionButtons({
+  slice: PropTypes.object,
+  addDangerToast: PropTypes.func.isRequired,
+  addSuccessToast: PropTypes.func.isRequired,
+};
```
Next time, if we are to update `propTypes`, might as well just convert the whole component into `tsx` first.

You could also just convert it to a functional component and use the `useToasts` hook.

michael-s-molina replied:
Updated to typescript!

ktmud commented on superset-frontend/src/dashboard/components/menu/ShareMenuItems.tsx:
```diff
@@ -0,0 +1,85 @@
+    ...rest
+  } = props;
+
+  const getShortUrl = useUrlShortener(url);
```
I think you can just let `useUrlShortener` return `shortUrl` itself and add a loading state if `shortUrl` is falsy.

michael-s-molina replied:
The idea here was correct but the implementation was not. The idea is that useUrlShortener returns an async function to delay the POST request to the moment the user actually clicks the option. As an async function you can infer loading and error states. I fixed the code to reflect this. Previously the error route (when the server throws an error) was not functioning and now is fixed also.

ktmud commented on superset-frontend/src/dashboard/components/menu/ShareMenuItems.tsx:
```diff
@@ -0,0 +1,85 @@
+      addSuccessToast(t('Copied to clipboard!'));
+    } catch (error) {
+      addDangerToast(
+        t('Sorry, your browser does not support copying. Use Ctrl / Cmd + C!'),
```
I don't think there is a place to `Ctrl + C` from after the updated UI.

michael-s-molina replied:
Great catch! Fixed.

ktmud commented on superset-frontend/src/dashboard/components/menu/ShareMenuItems.tsx:
```diff
@@ -0,0 +1,85 @@
+
+  async function onShareByEmail() {
+    try {
+      const bodyWithLink = t('%s%s', emailBody, getShortUrl());
```
`t` is probably not needed if this is just simple string concatenation.

michael-s-molina replied:
I checked that the parent component is responsible for the translation of email body. No need here. Removed.

junlincc commented:
@ktmud you wanna approve the code?

ktmud reviewed (approved):
Code LGTM with a couple of small comments.

A small UX suggestion: instead of using toast, you may also just updated the tooltip when copied was successful (you can even show a "Loading..." while fetching the shortened URL in the tooltip, too)

<img width="324" alt="copied-to-clipboard" src="https://user-images.githubusercontent.com/335541/108005424-8c5f1900-6fad-11eb-9cb6-4700e3fb20b3.png">

ktmud commented:
> @ktmud you wanna approve the code?

Was just reviewing and testing this locally. 😃

junlincc commented:
@michael-s-molina @ktmud Thank you both for perfecting this change! this is way better than my original expectation! We will still need to revisit the layout in the near future, but sure to keep the essence from this PR. ♥️

michael-s-molina commented on superset-frontend/spec/javascripts/dashboard/components/HeaderActionsDropdown_spec.jsx:
```diff
@@ -79,19 +79,19 @@ describe('HeaderActionsDropdown', () => {
     });
 
-    it('should render five Menu items', () => {
+    it('should render four Menu items', () => {
```
I changed to "should render available Menu items" since this is nested in a describe with the role being tested. This way future changes are not going to impact this text.

michael-s-molina commented:
> Code LGTM with a couple of small comments.
> 
> A small UX suggestion: instead of using toast, you may also just updated the tooltip when copied was successful (you can even show a "Loading..." while fetching the shortened URL in the tooltip, too)
> 
> <img alt="copied-to-clipboard" width="324" src="https://user-images.githubusercontent.com/335541/108005424-8c5f1900-6fad-11eb-9cb6-4700e3fb20b3.png">

Accepted you suggestion. Now we update the tooltip.

ktmud commented on superset-frontend/src/explore/components/ExploreActionButtons.tsx:
```diff
@@ -0,0 +1,204 @@
+  const getShortUrl = useUrlShortener(longUrl);
+
+  const doCopyLink = async () => {
+    try {
+      setCopyTooltip(t('Loading...'));
+      const shortUrl = await getShortUrl();
+      await copyTextToClipboard(shortUrl);
+      setCopyTooltip(t('Copied to clipboard!'));
+    } catch (error) {
+      setCopyTooltip(t('Sorry, your browser does not support copying.'));
+    }
```
```suggestion
    let shortUrl;
    try {
      setCopyTooltip(t('Loading...'));
      shortUrl = await getShortUrl();
    } catch (error) {
      setCopyTooltip(t('ERROR: failed to generate chart URL.'));
      return;
    }
    try {
      await copyTextToClipboard(shortUrl);
      setCopyTooltip(t('Copied to clipboard!'));
    } catch (error) {
      setCopyTooltip(t('Sorry, your browser does not support copying.'));
    }
```

I think we can use two try/catch to distinguish backend error from browser support error. For browsers that does not support copying, maybe you can also show the shortened URL in the tooltip (or an alert box, or a modal) so users can copy it themselves.

ktmud reviewed (approved):
One last comment, but let's do it in another PR (if you think it's necessary).
